paddle/fluid/extension/src/tensor.cc

```diff
@@ -85,129 +83,169 @@ DataType Tensor::type() const {
+  } else if (type == framework::proto::VarType::INT16) {
+    return DataType::INT16;
+  } else if (type == framework::proto::VarType::INT8) {
+    return DataType::INT8;
   } else if (type == framework::proto::VarType::UINT8) {
     return DataType::UINT8;
   } else if (type == framework::proto::VarType::FP64) {
     return DataType::FLOAT64;
+  } else if (type == framework::proto::VarType::BF16) {
+    return DataType::BFLOAT16;
+  } else if (type == framework::proto::VarType::FP16) {
+    return DataType::FLOAT16;
+  } else if (type == framework::proto::VarType::COMPLEX64) {
+    return DataType::COMPLEX64;
+  } else if (type == framework::proto::VarType::COMPLEX128) {
+    return DataType::COMPLEX128;
   }
   return DataType::FLOAT32;
 }
 
 template <typename T>
-void Tensor::copy_from_cpu(const T *data) {
+Tensor Tensor::copy_to_gpu() {
+#ifdef PADDLE_WITH_CUDA
   GET_CASTED_TENSOR;
   PADDLE_ENFORCE_GE(tensor->numel(), 0,
                     platform::errors::PreconditionNotMet(
                         "You should call Tensor::Reshape(const "
                         "std::vector<int> &shape)"
                         "function before copying data from cpu."));
   size_t ele_size = tensor->numel() * sizeof(T);
-
-  if (place_ == PlaceType::kCPU) {
-    auto *t_data = tensor->mutable_data<T>(platform::CPUPlace());
-    std::memcpy(static_cast<void *>(t_data), data, ele_size);
+  Tensor target = Tensor(PlaceType::kGPU);
+  target.reshape(shape());
+  auto *p_target_data = target.template mutable_data<T>();
+  auto p_src_data = tensor->data<T>();
+
+  platform::DeviceContextPool &pool = platform::DeviceContextPool::Instance();
+  int device_num = platform::GetCurrentDeviceId();
+  platform::CUDAPlace gpu_place(device_num);
+  auto *dev_ctx =
+      static_cast<const platform::CUDADeviceContext *>(pool.Get(gpu_place));
+  if (platform::is_cpu_place(tensor->place())) {
+    memory::Copy(gpu_place, static_cast<void *>(p_target_data),
+                 platform::CPUPlace(), p_src_data, ele_size, dev_ctx->stream());
   } else {
-#ifdef PADDLE_WITH_CUDA
-    platform::DeviceContextPool &pool = platform::DeviceContextPool::Instance();
-    int device_num = platform::GetCurrentDeviceId();
-    platform::CUDAPlace gpu_place(device_num);
-    auto *t_data = tensor->mutable_data<T>(gpu_place);
-    auto *dev_ctx =
-        static_cast<const platform::CUDADeviceContext *>(pool.Get(gpu_place));
-
-    memory::Copy(gpu_place, static_cast<void *>(t_data), platform::CPUPlace(),
-                 data, ele_size, dev_ctx->stream());
+    memory::Copy(gpu_place, static_cast<void *>(p_target_data), gpu_place,
+                 p_src_data, ele_size, dev_ctx->stream());
+  }
+  cudaStreamSynchronize(dev_ctx->stream());
+  return target;
 #else
-    PADDLE_THROW(platform::errors::Unavailable(
-        "Not compiled with CUDA, should not reach here."));
+  PADDLE_THROW(platform::errors::Unavailable(
+      "Not compiled with CUDA, should not reach here."));
```

Comment:
polish error message, `Paddle is not compiled with CUDA.` is ok, `should not` is not needed

Reply:
done